test/utils/server.js

```diff
@@ -1,73 +1,47 @@
-// no forking, run just one process when testing
-config.conf.num_workers = 0;
-// have a separate, in-memory logger only
-config.conf.logging = {
-    name: 'test-log',
-    level: 'trace',
-    stream: logStream()
-};
-// make a deep copy of it for later reference
-const origConfig = extend(true, {}, config);
-
-module.exports.stop = () => { return BBPromise.resolve(); };
-let options = null;
-const runner = new ServiceRunner();
-
-
-function start(_options) {
-
-    _options = _options || {};
+class TestServiceTemplateNodeRunner extends TestRunner {
+    constructor(configPath = `${__dirname}/../../config.yaml`) {
+        super(configPath);
+        this._spec = null;
+    }
 
-    if (!assert.isDeepEqual(options, _options)) {
-        console.log('starting test server'); // eslint-disable-line no-console
-        return module.exports.stop().then(() => {
-            options = _options;
-            // set up the config
-            config = extend(true, {}, origConfig);
-            extend(true, config.conf.services[myServiceIdx].conf, options);
-            return runner.start(config.conf)
-            .then((serviceReturns) => {
-                module.exports.stop = () => {
-                    console.log('stopping test server'); // eslint-disable-line no-console
-                    serviceReturns.forEach(servers =>
-                        servers.forEach(server =>
-                            server.shutdown()));
-                    return runner.stop().then(() => {
-                        module.exports.stop = () => {
-                            return BBPromise.resolve();
-                        };
-                    });
-                };
-                return true;
+    get config() {
+        if (!this._running) {
+            throw new Error('Accessing test service config before starting the service');
+        }
+
+        // build the API endpoint URI by supposing the actual service
+        // is the last one in the 'services' list in the config file
+        const myServiceIdx = this._runner._impl.config.services.length - 1;
+        const myService = this._runner._impl.config.services[myServiceIdx];
+        const uri = `http://localhost:${myService.conf.port}/`;
+        if (!this._spec) {
+            // We only want to load this once.
+            preq.get(`${uri}?spec`)
+            .then((res) => {
+                if (!res.body) {
+                    throw new Error('Failed to get spec');
+                }
+                // save a copy
+                this._spec = res.body;
+            })
+            .catch((err) => {
+                // this error will be detected later, so ignore it
+                this._spec = { paths: {}, 'x-default-params': {} };
             });
-        });
-    } else {
-        return BBPromise.resolve();
+        }
+
+        return {
+            uri,
+            service: myService,
+            conf: this._runner._impl.config,
+            spec: this._spec
```

Comment:
If the condition on L22 is true (i.e. there's no spec loaded yet), this will be null. You need to return thi sonly once the request promise has been fulfilled.

Reply:
Good point. Changed it so that the promise returns the value when completed otherwise the saved spec is used.

Reply:
Euh, I don't see this in the code. It's still got a straight return on L46 regardless of whether the spec was present or not...